### ISSUE TITLE
TEST-55-OOMD occasionally fails on Ubuntu CI

### ISSUE THREAD

**Issue by bluca** [systemd/systemd#28504]
eg: https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-upstream-systemd-ci-systemd-ci/jammy/amd64/s/systemd-upstream/20230721_061305_4c0c2@/log.gz
from: https://github.com/systemd/systemd/pull/28479

```
13647s [  196.603727] testsuite-55.sh[498]: ○ testsuite-55-testchill.service - No memory pressure
13647s [  196.730527] testsuite-55.sh[498]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-te
stchill.service; static)
13647s [  196.851489] testsuite-55.sh[498]:      Active: inactive (dead)
```

**Comment by mrc0mmand** [comment]
I'm going through logs and trying to reproduce this, and while doing so I noticed a couple of warnings/errors (some of them we even discussed on Telegram iirc):

```
[   29.173407] H testsuite-55.sh[664]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
[   29.210845] H systemd[1]: Started run-u50.service.
[   29.224695] H systemd[1]: Started session-12.scope.
[   29.226876] H (o-bridge)[893]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.234523] H testsuite-55.sh[891]: × testsuite-55-testbloat.service - Create a lot of memory pressure
[   29.234523] H testsuite-55.sh[891]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
[   29.234523] H testsuite-55.sh[891]:      Active: failed (Result: oom-kill) since Mon 2023-07-24 22:05:47 CEST; 950ms ago
[   29.234523] H testsuite-55.sh[891]:    Duration: 13.720s
[   29.234523] H testsuite-55.sh[891]:     Process: 787 ExecStart=stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1 (code=killed, signal=KILL)
[   29.234523] H testsuite-55.sh[891]:    Main PID: 787 (code=killed, signal=KILL)
[   29.234523] H testsuite-55.sh[891]:         CPU: 270ms
[   29.246102] H dbus-broker[777]: Peer :1.12 is being disconnected as it sent a message with an invalid header.
[   29.246429] H (sd-pam)[894]: pam_systemd(login:session): Failed to release session: Access denied
[   29.246720] H testsuite-55.sh[664]: + break
[   29.246720] H testsuite-55.sh[664]: + systemctl --machine testuser@.host --user status testsuite-55-testbloat.service
[   29.249797] H (sd-pam)[894]: pam_unix(login:session): session closed for user testuser
[   29.249949] H (sd-pam)[894]: PAM Attempted to close sd-bus after fork, this should not happen.
[   29.250586] H systemd[1]: run-u50.service: Deactivated successfully.
[   29.250866] H dbus-broker[650]: A security policy denied :1.52 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
[   29.254817] H systemd[1]: session-12.scope: Deactivated successfully.
[   29.279868] H systemd[1]: Started run-u53.service.
[   29.295880] H systemd[1]: Started session-13.scope.
[   29.303401] H (o-bridge)[900]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.313469] H dbus-broker[777]: Peer :1.13 is being disconnected as it sent a message with an invalid header.
[   29.313728] H testsuite-55.sh[895]: × testsuite-55-testbloat.service - Create a lot of memory pressure
[   29.313728] H testsuite-55.sh[895]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testbloat.service; static)
[   29.313728] H testsuite-55.sh[895]:      Active: failed (Result: oom-kill) since Mon 2023-07-24 22:05:47 CEST; 1s ago
[   29.313728] H testsuite-55.sh[895]:    Duration: 13.720s
[   29.313728] H testsuite-55.sh[895]:     Process: 787 ExecStart=stress --timeout 3m --vm 10 --vm-bytes 200M --vm-keep --vm-stride 1 (code=killed, signal=KILL)
[   29.313728] H testsuite-55.sh[895]:    Main PID: 787 (code=killed, signal=KILL)
[   29.313728] H testsuite-55.sh[895]:         CPU: 270ms
[   29.323777] H (sd-pam)[901]: pam_systemd(login:session): Failed to release session: Access denied
[   29.324036] H testsuite-55.sh[664]: + systemctl --machine testuser@.host --user status testsuite-55-testchill.service
[   29.325619] H (sd-pam)[901]: pam_unix(login:session): session closed for user testuser
[   29.325743] H (sd-pam)[901]: PAM Attempted to close sd-bus after fork, this should not happen.
[   29.326192] H systemd[1]: run-u53.service: Deactivated successfully.
[   29.326425] H dbus-broker[650]: A security policy denied :1.55 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
[   29.331144] H systemd[1]: session-13.scope: Deactivated successfully.
[   29.353609] H systemd[1]: Started run-u56.service.
[   29.385704] H systemd[1]: Started session-14.scope.
[   29.395198] H (o-bridge)[907]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[   29.407198] H dbus-broker[777]: Peer :1.14 is being disconnected as it sent a message with an invalid header.
```

There are three possible issues:

```
[   29.246102] H dbus-broker[777]: Peer :1.12 is being disconnected as it sent a message with an invalid header.
[   29.250866] H dbus-broker[650]: A security policy denied :1.52 to send method call /org/freedesktop/login1:org.freedesktop.login1.Manager.ReleaseSession to org.freedesktop.login1.
...
[   29.246429] H (sd-pam)[894]: pam_systemd(login:session): Failed to release session: Access denied
...
[   29.249949] H (sd-pam)[894]: PAM Attempted to close sd-bus after fork, this should not happen.
```

Are any of these something _we_ should/can fix on our side?

**Comment by bluca** [comment]
the last one means we are missing some cleanup in some error path

**Comment by mrc0mmand** [comment]
And to the original issue - something stops the user systemd daemon during the test, which also stops the `testchill` service:

```
[  182.230456] H testsuite-55.sh[364]: + systemctl --machine testuser@.host --user status testsuite-55-testchill.service
[  184.409685] H systemd[1]: Started systemd-stdio-bridge -punix:path=${XDG_RUNTIME_DIR}/bus.
[  184.951159] H (o-bridge)[500]: pam_env(login:session): Unable to open env file: /etc/default/locale: No such file or directory
[  184.981360] H (o-bridge)[500]: pam_lastlog(login:session): conversation failed
[  184.999406] H (o-bridge)[500]: pam_unix(login:session): session opened for user testuser(uid=4711) by (uid=0)
[  185.463902] H systemd[1]: Stopping User Manager for UID 4711...
[  185.516932] H systemd[352]: Activating special unit Exit the Session...
[  185.623859] H systemd[352]: Stopped target Main User Target.
[  185.646005] H systemd[352]: Stopped target Timers.
[  185.735610] H systemd[352]: Stopping D-Bus User Message Bus...
[  185.773351] H systemd[352]: Stopping No memory pressure...
[  185.959869] H systemd[352]: Stopped No memory pressure.
[  186.014400] H systemd[352]: Stopped D-Bus User Message Bus.
[  186.222129] H systemd[352]: Removed slice Test slice for memory pressure kills.
[  186.223224] H systemd[352]: testsuite-55-workload.slice: Consumed 2.375s CPU time.
[  186.239196] H systemd[352]: Removed slice Slice /testsuite/55.
[  186.256806] H systemd[352]: testsuite-55.slice: Consumed 2.375s CPU time.
[  186.261682] H systemd[352]: Removed slice Slice /testsuite.
[  186.262957] H systemd[352]: testsuite.slice: Consumed 2.375s CPU time.
[  186.263890] H systemd[352]: Stopped target Basic System.
[  186.265562] H systemd[352]: Stopped target Paths.
[  186.266541] H systemd[352]: Stopped target Sockets.
[  186.267488] H systemd[352]: Stopped target System Initialization.
[  186.301042] H systemd[352]: Closed D-Bus User Message Bus Socket.
[  186.302135] H systemd[352]: Removed slice User Application Slice.
[  186.310315] H systemd[352]: Reached target Shutdown.
[  186.311798] H systemd[352]: Finished Exit the Session.
[  186.317628] H systemd[352]: Reached target Exit the Session.
[  186.627969] H systemd[1]: user@4711.service: Deactivated successfully.
[  186.667464] H systemd[1]: Stopped User Manager for UID 4711.
[  186.690304] H systemd[1]: user@4711.service: Consumed 9.791s CPU time.
[  186.977139] H systemd[1]: Stopping User Runtime Directory /run/user/4711...
[  189.058944] H (systemd)[508]: pam_unix(systemd-user:session): session opened for user testuser(uid=4711) by (uid=0)
[  189.249834] H systemd[1]: run-user-4711.mount: Deactivated successfully.
[  189.290185] H systemd[1]: user-runtime-dir@4711.service: Deactivated successfully.
[  189.382396] H systemd[1]: Starting User Runtime Directory /run/user/4711...
[  189.477233] H systemd[1]: Finished User Runtime Directory /run/user/4711.
[  189.518378] H systemd[1]: Starting User Manager for UID 4711...
[  194.189319] H systemd[508]: Queued start job for default target Main User Target.
[  194.315115] H systemd[508]: Created slice User Application Slice.
[  194.321588] H systemd[508]: Reached target Paths.
[  194.324803] H systemd[508]: Reached target System Initialization.
[  194.328668] H systemd[508]: Reached target Timers.
[  194.377569] H systemd[508]: Starting D-Bus User Message Bus Socket...
[  194.803634] H systemd[508]: Listening on D-Bus User Message Bus Socket.
[  194.811434] H systemd[508]: Reached target Sockets.
[  194.814635] H systemd[508]: Reached target Basic System.
[  194.844513] H systemd[508]: Reached target Main User Target.
[  194.846899] H systemd[1]: Started User Manager for UID 4711.
[  194.852680] H systemd[508]: Startup finished in 5.216s.
[  195.001466] H systemd[1]: Started Session 4 of User testuser.
[  195.690475] H systemd[508]: Started D-Bus User Message Bus.
[  196.111658] H dbus-daemon[518]: [session uid=4711 pid=518] AppArmor D-Bus mediation is enabled
[  196.566922] H systemd-stdio-bridge[500]: Failed to send message: Transport endpoint is not connected
[  196.585623] H testsuite-55.sh[498]: ○ testsuite-55-testchill.service - No memory pressure
[  196.585623] H testsuite-55.sh[498]:      Loaded: loaded (/usr/lib/systemd/tests/testdata/units/testsuite-55-testchill.service; static)
[  196.585623] H testsuite-55.sh[498]:      Active: inactive (dead)
```

Not sure if it's related to the two errors it shows in between:

```
[  184.951159] H (o-bridge)[500]: pam_env(login:session): Unable to open env file: /etc/default/locale: No such file or directory
[  184.981360] H (o-bridge)[500]: pam_lastlog(login:session): conversation failed
```

that show up a couple of times throughout the test.

**Comment by bluca** [comment]
I see those (and the ones above) in a successful run too

**Comment by mrc0mmand** [comment]
Do you have a journal from a successful run for comparison? All I can currently see is that apparently logind enqueues a stop job for the user session service, but I'm not sure why. Also, I don't see any of the mentioned errors/warnings on my Fedora machine, so I have no idea which one might be a culprit.

```
[   96.585495] H systemd[1]: systemd-logind.service: D-Bus name org.freedesktop.login1 now owned by :1.2
...
[  185.110662] H systemd[1]: Got message type=method_call sender=:1.2 destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StopUnit  cookie=67 reply_cookie=0 signature=ss error-name=n/a error-message=n/a
[  185.117007] H systemd[1]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=GetConnectionUnixUser cookie=1565 reply_cookie=0 signature=s error-name=n/a error-message=n/a
[  185.118928] H systemd[1]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.1 path=n/a interface=n/a member=n/a  cookie=54 reply_cookie=1565 signature=u error-name=n/a error-message=n/a
[  185.119888] H systemd[1]: user@4711.service: Trying to enqueue job user@4711.service/stop/fail
[  185.145731] H systemd[1]: Added job user@4711.service/stop to transaction.
[  185.146795] H systemd[1]: user@4711.service: Installed new job user@4711.service/stop as 634
[  185.147732] H systemd[1]: user@4711.service: Enqueued job user@4711.service/stop as 634
```

**Comment by bluca** [comment]
[good.tar.gz](https://github.com/systemd/systemd/files/12161485/good.tar.gz)

here it is, you can see the same pam errors

**Comment by mrc0mmand** [comment]
> And to the original issue - something stops the user systemd daemon during the test, which also stops the `testchill` service:

I apparently didn't connect all the dots at the time I wrote that - the root cause here is that the test user didn't have session lingering turned on. This should be resolved by #31431.
